mule-shaders-module/src/main/java/org/mule/module/shaders/api/ReflectiveAccessShader.java

```diff
@@ -0,0 +1,632 @@
+import org.apache.maven.plugins.shade.resource.ReproducibleResourceTransformer;
+import org.apache.maven.plugins.shade.resource.ResourceTransformer;
+import org.codehaus.plexus.component.annotations.Component;
+import org.codehaus.plexus.logging.AbstractLogEnabled;
+import org.codehaus.plexus.util.IOUtil;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.commons.Remapper;
+
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.io.PushbackInputStream;
+import java.io.Writer;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.jar.JarOutputStream;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.zip.CRC32;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipException;
+
+/**
+ *  A shader capable of changing the specified classes for reflective access before relocating packages.
+ *  Custom version of {@link org.apache.maven.plugins.shade.DefaultShader}.
+ */
+@Component(role = Shader.class, hint = "ReflectiveAccessShader")
+public class ReflectiveAccessShader
+    extends AbstractLogEnabled
+    implements Shader {
+
+  private static final int BUFFER_SIZE = 32 * 1024;
+
+  public void shade(ShadeRequest shadeRequest)
+      throws IOException, MojoExecutionException {
+    Set<String> resources = new HashSet<>();
+
+    ManifestResourceTransformer manifestTransformer = null;
+    List<ResourceTransformer> transformers =
+        new ArrayList<>(shadeRequest.getResourceTransformers());
+    for (Iterator<ResourceTransformer> it = transformers.iterator(); it.hasNext();) {
+      ResourceTransformer transformer = it.next();
+      if (transformer instanceof ManifestResourceTransformer) {
+        manifestTransformer = (ManifestResourceTransformer) transformer;
+        it.remove();
+      }
+    }
+
+    RelocatorRemapper remapper = new RelocatorRemapper(shadeRequest.getRelocators());
+
+    // noinspection ResultOfMethodCallIgnored
+    shadeRequest.getUberJar().getParentFile().mkdirs();
+
+    JarOutputStream out = null;
+    try {
+      out = new JarOutputStream(new BufferedOutputStream(new FileOutputStream(shadeRequest.getUberJar())));
+      goThroughAllJarEntriesForManifestTransformer(shadeRequest, resources, manifestTransformer, out);
+
+      // CHECKSTYLE_OFF: MagicNumber
+      Multimap<String, File> duplicates = HashMultimap.create(10000, 3);
+      // CHECKSTYLE_ON: MagicNumber
```

Comment:
Consider just defining the constant for these numbers. Giving an explicit name to them would make more clear the purpose of the value.

Reply:
Unfortunately, there was no way to extend the original shader code from `org.apache.maven.plugins.shade.DefaultShader` in order to introduce the custom remapping code that we need, so the implementation was just copied over.

Only one line of code needed to be changed from the original implementation:

```diff
// Original line:
--- ClassVisitor cv = new ClassRemapper(cw, remapper) {
// Modified line:
+++ ClassVisitor cv = new ReflectiveAccessClassRemapper(cw, remapper, transformers) {
```

There are plenty of improvements that can be made over the original default shader code, but it would be easier to keep track of customisations if the changes to it are minimal.